sdcm/utils/k8s.py

```diff
@@ -609,6 +614,251 @@ def stop(self, timeout=None):
         self.join(timeout)
 
 
+class ScyllaPodsIPChangeTrackerThread(threading.Thread):
+    """It tracks changes of Scylla pods IP addresses in whole K8S cluster and updates mapping.
+
+    When a new IP is detected ...
+    - the mapping gets updated
+    - 'info' log message gets created
+    - 'info' event gets created.
+    - the monitoring node related to a Scylla pod gets reconfigured.
+
+    'mapper_dict' has following structure:
+    {
+        'scylla': {
+            'sct-cluster-dc-1-kind-0': {
+                'current_ip': '10.0.0.4',
+                'old_ips': ['10.0.0.2', '10.0.0.3'],
+            },
+            ...
+        },
+        'scylla-manager': {
+            'scylla-manager-manager-dc-manager-rack-0': {
+                'current_ip': '10.0.2.2',
+                'old_ips': [],
+            },
+            ...
+        },
+        ...
+    }
+    """
+
+    SCYLLA_PODS_SELECTOR = "app.kubernetes.io/name=scylla"
+    SCYLLA_PODS_EXPECTED_LABEL_KEYS = ("scylla/datacenter", "scylla/rack")
+
+    def __init__(self, k8s_kluster, mapper_dict):
+        self._termination_event = threading.Event()
+        super().__init__(daemon=True, name=self.__class__.name)
+        self.k8s_kluster = k8s_kluster
+        self.mapper_dict = mapper_dict
+        self.watcher = None
+        self._k8s_core_v1_api = KubernetesOps.core_v1_api(self.k8s_kluster.get_api_client())
+
+    @retrying(n=3600, sleep_time=1, allowed_exceptions=(ConnectionError, ))
+    def _open_stream(self, cache={}) -> None:  # pylint: disable=dangerous-default-value
+        try:
+            now = time.time()
+            if cache.get("last_call_at", 0) + 5 > now:
+                # NOTE: sleep for some time if it is called too often
+                time.sleep(5)
+                now = time.time()
+            cache["last_call_at"] = now
+
+            self.watcher = self._k8s_core_v1_api.list_endpoints_for_all_namespaces(
+                label_selector=self.SCYLLA_PODS_SELECTOR,
+                watch=True,
+                async_req=False,
+                _preload_content=False)
+        except k8s.client.rest.ApiException as exc:
+            LOGGER.warning("'_open_stream()': failed to open stream:\n%s", exc)
+            # NOTE: following is workaround for the error 401 which may happen due to
+            #       some config data corruption during the forced socket connection failure
+            self._k8s_core_v1_api = KubernetesOps.core_v1_api(self.k8s_kluster.get_api_client())
+            raise ConnectionError(str(exc)) from None
+
+    @retrying(n=3600, sleep_time=1,  allowed_exceptions=(
+        ProtocolError, IncompleteRead, ReadTimeoutError, TimeoutError, ValueError))
+    def _read_stream(self) -> Iterator[str]:
+        while True:
```

Comment:
how this thread is being stopped ? if we are in an endless loop ?

Reply:
At first, we haven't been stopping any of the threads running in K8S. Why? Because we call this code in the "cleanup" stage where it is not initiated. So, all the additional threads we have are not stopped explicitly.
At second, we must not stop this thread somewhere in the middle. And it is totally ok to be closed automatically with the main python process.

In a newer version I have in my dev branch I added `_termination_event` attribute, but it is not really needed.
Due to the nature of this thread, it will hang. It will hang due to the endless timeout which we need because it is needed by the design - `wait for updates`.

Reply:
What happen when we'll need to handle multiple cluster, I still think we need ability to stop this